iopipe/contrib/trace/auto_http.py

```diff
@@ -0,0 +1,137 @@
+
+if Session is not None:
+    original_session_send = Session.send
+
+if BotocoreSession is not None:
+    original_botocore_session_send = BotocoreSession.send
+
+EXCLUDE_HEADERS = ['Authorization', 'Cookie', 'Proxy-Authorization', 'Set-Cookie']
+
+REQUEST_KEYS = [
+    ('method', 'method'),
+    ('url', 'url'),
+]
+
+RESPONSE_KEYS = [
+    ('status_code', 'statusCode'),
+]
+
+URLPARSE_KEYS = [
+    ('fragment', 'hash'),
+    ('hostname', 'hostname'),
+    ('path', 'path'),
+    ('port', 'port'),
+    ('query', 'query'),
+    ('scheme', 'protocol'),
+]
+
+
+def patch_session_send(context, auto_measure, http_filter):
+    """
+    Monkey patches requests' Session class, if available. Overloads the
+    send method to add tracing and metrics collection.
+    """
+    if Session is None:
+        return
+
+    def send(self, *args, **kwargs):
+        id = str(uuid.uuid4())
+        with context.iopipe.mark(id):
+            response = original_session_send(self, *args, **kwargs)
+        if not auto_measure:
+            context.iopipe.mark.measure(id)
+        collect_metrics_for_response(response, context, id, http_filter)
+        return response
+
+    Session.send = send
+
+
+def patch_botocore_session_send(context, auto_measure, http_filter):
+    """
+    Monkey patches botocore's vendored requests, if available. Overloads the
+    Session class' send method to add tracing and metric collection.
+    """
+    if BotocoreSession is None:
+        return
+
+    def send(self, *args, **kwargs):
+        id = str(uuid.uuid4())
+        with context.iopipe.mark(id):
+            response = original_botocore_session_send(self, *args, **kwargs)
+        if not auto_measure:
+            context.iopipe.mark.measure(id)
+        collect_metrics_for_response(response, context, id, http_filter)
+        return response
+
+    BotocoreSession.send = send
+
+
+def restore_session_send():
+    """Restores the original Session send method"""
+    if Session is not None:
+        Session.send = original_session_send
+
+
+def restore_botocore_session_send():
+    """Restores the original botocore Session send method"""
+    if BotocoreSession is not None:
+        BotocoreSession.send = original_botocore_session_send
+
+
+def patch_requests(context, auto_measure, http_filter):
+    patch_session_send(context, auto_measure, http_filter)
+    patch_botocore_session_send(context, auto_measure, http_filter)
+
+
+def restore_requests():
+    restore_session_send()
+    restore_botocore_session_send()
+
+
+def collect_metrics_for_response(response, context, id, http_filter):
+    """
+    Collects relevant metrics from a requests Response object and adds them to
+    the IOpipe context.
+    """
+    if http_filter is not None and callable(http_filter):
+        response = http_filter(response)
+        if response is False:
+            context.iopipe.mark.delete(id)
+            return
+
+    prefix = functools.partial('@iopipe/trace.{}.{}'.format, id)
```

Comment:
Why store in custom-metrics instead of in `performanceEntries` with other tracing data?

Reply:
We do store the traces themselves in `performanceEntries`, but the `performanceEntries` schema doesn't permit key/value data like that which we collect for an HTTP request. As such, we store all the HTTP request data in `custom_metrics` as that is the only place currently that will allow for key/value data within the schema. It is similar to what we do with `@iopipe/event-info`. We've discussed separate fields for event info and auto traces, but both amount to key/value data.

Reply:
Thank you, please make sure that's noted in the wiki for this then, since that appears to be part of the corroborating flow